tsl/src/chunk.c

```diff
@@ -186,3 +206,889 @@ chunk_invoke_drop_chunks(Oid relid, Datum older_than, Datum older_than_type, boo
+	/* Step 1: Do sanity checks and then prepare to sort rels in consistent order. */
+	for (int i = 0; i < nrelids; i++)
+	{
+		Oid relid = DatumGetObjectId(relids[i]);
+		char relpers = '\0';
+		const Chunk *chunk;
+		Relation rel;
+		Oid amoid;
+
+		if (nulls[i] || !OidIsValid(relid))
+			ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE), errmsg("invalid relation")));
+
+		if (i > 0 && DatumGetObjectId(relids[i]) == DatumGetObjectId(relids[i - 1]))
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("duplicate relation \"%s\" in merge",
+							get_rel_name(DatumGetObjectId(relids[i])))));
+
+		/* Lock the relation before doing other checks that can lock dependent
+		 * objects (this can otherwise lead to deadlocks with concurrent
+		 * operations). Note that if we take ExclusiveLock here to allow
+		 * readers while we are rewriting/merging the relations, the lock
+		 * needs to be upgraded to an AccessExclusiveLock later. This can also
+		 * lead to deadlocks.
+		 *
+		 * Ideally, we should probably take locks on all dependent objects as
+		 * well, at least on chunk-related objects that will be
+		 * dropped. Otherwise, that might also cause deadlocks later. For
+		 * example, if doing a concurrent DROP TABLE on one of the chunks will
+		 * lead to deadlock because it grabs locks on all dependencies before
+		 * dropping.
+		 *
+		 * However, for now we won't do that because that requires scanning
+		 * pg_depends and concurrent operations will probably fail anyway if
+		 * we remove the objects. We might as well fail with a deadlock.
+		 */
+		LOCKMODE lockmode =
+			(lock_upgrade == MERGE_LOCK_ACCESS_EXCLUSIVE) ? AccessExclusiveLock : ExclusiveLock;
+
+		rel = try_table_open(relid, lockmode);
+
+		/* Check if the table actually exists. If not, it could have been
+		 * deleted in a concurrent merge. */
+		if (rel == NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_TABLE),
+					 errmsg("relation does not exist"),
+					 errdetail("The relation with OID %u might have been removed "
+							   "by a concurrent merge or other operation.",
+							   relid)));
+
+		if (rel->rd_rel->relkind == '\0')
+			ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_TABLE),
+					 errmsg("relation with OID %u does not exist", relid)));
+
+		if (rel->rd_rel->relkind != RELKIND_RELATION)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("cannot merge non-table relations")));
+
+		/* Only owner is allowed to merge */
+		if (!object_ownercheck(RelationRelationId, relid, GetUserId()))
+			aclcheck_error(ACLCHECK_NOT_OWNER,
+						   get_relkind_objtype(rel->rd_rel->relkind),
+						   get_rel_name(relid));
+
+		/* Remember the relpersistence until processing next rel */
+		relpersistence = relpers;
+
+		if (relpersistence != '\0' && relpersistence != relpers)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("cannot merge relations with different persistence")));
```

Comment:
Do we really need those relpersistence checks since we don't support unlogged hypertables/chunks?

Reply:
I guess for now we don't, so I could remove it. The check was included in some PG code I based this on so I included it for completeness.